src/dispatch/stage.rs

```diff
@@ -0,0 +1,438 @@
+    reads: Vec<GroupVec<SmallVec<[ResourceId; 12]>>>,
+    running_time: Vec<GroupVec<u8>>,
+    stages: Vec<Stage<'a>>,
+    writes: Vec<GroupVec<SmallVec<[ResourceId; 10]>>>,
+}
+
+impl<'a> StagesBuilder<'a> {
+    pub fn add_barrier(&mut self) {
+        self.barrier = self.stages.len();
+    }
+
+    pub fn insert<T>(&mut self, mut dep: SmallVec<[SystemId; 4]>, id: SystemId, system: T)
+        where T: for<'b> System<'b> + Send + 'a
+    {
+        use system::SystemData;
+
+        let mut reads = T::SystemData::reads(0);
+        let writes = T::SystemData::writes(0);
+
+        reads.sort();
+        reads.dedup();
+
+        let new_time = system.running_time();
+
+        let target = self.insertion_target(&reads, &writes, &mut dep, new_time);
+
+        let (stage, group) = match target {
+            InsertionTarget::Stage(stage) => {
+                let group = self.ids[stage].len();
+                self.add_group(stage);
+
+                (stage, group)
+            }
+            InsertionTarget::Group(stage, group) => (stage, group),
+            InsertionTarget::NewStage => {
+                let stage = self.stages.len();
+
+                self.add_stage();
+                self.add_group(stage);
+
+                (stage, 0)
+            }
+        };
+
+        self.ids[stage][group].push(id);
+        self.reads[stage][group].extend(reads);
+        self.running_time[stage][group] += new_time as u8;
+        self.stages[stage].groups[group].push(Box::new(system));
+        self.writes[stage][group].extend(writes);
+    }
+
+    pub fn build(self) -> Vec<Stage<'a>> {
+        self.stages
+    }
+
+    fn add_stage(&mut self) {
+        self.ids.push(GroupVec::new());
+        self.reads.push(GroupVec::new());
+        self.running_time.push(GroupVec::new());
+        self.stages.push(Stage::new());
+        self.writes.push(GroupVec::new());
+    }
+
+    fn add_group(&mut self, stage: usize) {
+        self.ids[stage].push(ArrayVec::new());
+        self.reads[stage].push(SmallVec::new());
+        self.running_time[stage].push(0);
+        self.stages[stage].groups.push(ArrayVec::new());
+        self.writes[stage].push(SmallVec::new());
+    }
+
+    fn insertion_target<'rw, R, W>(&self,
+                                   new_reads: R,
+                                   new_writes: W,
+                                   new_dep: &mut SmallVec<[SystemId; 4]>,
+                                   new_time: RunningTime)
+                                   -> InsertionTarget
+        where R: IntoIterator<Item = &'rw ResourceId>,
+              R::IntoIter: Clone,
+              W: IntoIterator<Item = &'rw ResourceId>,
+              W::IntoIter: Clone
+    {
+        let new_reads = new_reads.into_iter();
+        let new_writes = new_writes.into_iter();
+
+        (self.barrier..self.stages.len())
+            .map(|stage| {
+                let conflict = Self::find_conflict(&*self.ids,
```

Comment:
any reason this is not just `self.find_conflict(stage, new_reads.clone(), new_writes.clone(), new_dep)`?